Fix embedded videos got removed

RadhiFadlillah opened the pull request:
This PR should fix #410.

Previously, in [`_cleanConditionally`](https://github.com/mozilla/readability/blob/master/Readability.js#L1619), we only check the `embed` tag for video container. However, the embedded videos usually found inside `iframe` or maybe `object`. In fact, that's how [`_clean`](https://github.com/mozilla/readability/blob/master/Readability.js#L1479) method filters which embed that must be deleted.

This PR add a simple check to see if the `node` is actually a video container. I've also added two tests from URLs provided in #410. If you think it's too much, feel free to remove it.

gijsk commented on Readability.js:
```diff
@@ -1661,10 +1661,27 @@ Readability.prototype = {
+          if (this.REGEXPS.videos.test(attributeValues) || this.REGEXPS.videos.test(node.innerHTML)) {
+            videoCount += 1;
+          } else {
             embedCount += 1;
```
So now we're counting objects and iframes as embeds as well, which will cause the container to be removed when there are more than 1 (see checks below this code). I think that's probably fine but it's worth flagging up.

RadhiFadlillah replied:
I'm not sure if I follow. So, you mean we need to add new flag to specify whether we will :

- only count the `embed` tags (like before PR); or 
- count all `object`, `iframe` and `embed` (as in this PR) ?

If so, how should I name the flag ? Will `FLAG_COUNT_EMBED_ONLY` works ?

gijsk replied:
I'm just noting that there's a side-effect to your patch that is quite different from the main goal of the patch, namely that if there's an arbitrary container node which contains 2 iframes that don't point to videos, the container will be removed.

I think that that side effect is probably OK; I don't think we need a separate flag for it. If we want to be more cautious, we could only increment `embedCount` if the node is really an `embed` node (ie not increment it for the iframe/object case).

Either way though, we don't need `videoCount`. If there is any video, we can return false immediately and stop all the loops. Because we never decrement `videoCount`, so when we return, for sure `videoCount` is still > 0, so we pass the condition below the loop:

```
        // If contains video, don't delete this node
        if (videoCount > 0) {
          return false;
        }
```

RadhiFadlillah replied:
> Either way though, we don't need `videoCount`

I've done it in [0a23f5](https://github.com/mozilla/readability/pull/526/commits/0a23f536326a5af02b7b10f5771cddfa82385217). Now, if the node contains an embedded video, the process will return false immediately and the `node` won't be deleted.

RadhiFadlillah commented on Readability.js:
```diff
@@ -1661,10 +1661,27 @@ Readability.prototype = {
+            return attr.value;
+          }).join("|");
+
+          if (this.REGEXPS.videos.test(attributeValues) || this.REGEXPS.videos.test(node.innerHTML)) {
```
That might be true. I just do it following the [`_clean`](https://github.com/mozilla/readability/blob/9b8ce0b3545a48f1a66079fc6855671d30105682/Readability.js#L1479) method.

I don't have many experiences on embedding video, so I just followed the existing code. Should I remove it and only check the attribute ?

gijsk replied:
Yeah, let's check only the attribute for now. :-)

gijsk replied:
Yeah, I don't think we need to to check innerHTML - let's check only the attributes and we'll see if anything breaks. :-)

gijsk replied:
Ah, hm, perhaps we need to check `innerHTML` in the `<object>` case - except that I don't think we actually use `<object>` for video embedding these days, only when it was still flash...

RadhiFadlillah replied:
Done in [0a23f5](https://github.com/mozilla/readability/pull/526/commits/0a23f536326a5af02b7b10f5771cddfa82385217), now `innerHTML` only checked if embed is an `<object>`.

RadhiFadlillah commented on Readability.js:
```diff
@@ -1661,10 +1661,27 @@ Readability.prototype = {
+
         for (var ei = 0, il = embeds.length; ei < il; ei += 1) {
-          if (!this.REGEXPS.videos.test(embeds[ei].src))
+          var attributeValues = [].map.call(node.attributes, function(attr) {
```
For this one, I also followed code in [`_clean`](https://github.com/mozilla/readability/blob/9b8ce0b3545a48f1a66079fc6855671d30105682/Readability.js#L1484) method as well.

I wonder which one is more expensive :

- running regex once for a long string, or
- running regex multiple times for a short string.

I'm more experienced in Go than JS, and in Go (AFAIK) running regex once is cheaper than running regex multiple times. With that said, I thought JS is the same in that aspect, which is why `_clean` method do it like this.

So, what do you think ? Should I make it loop over the attributes, or keep it like this ?

gijsk replied:
It depends which is the common case. Here, we have a list of embeds/iframes/objects. Really, we should be checking the actual `src` attributes instead of blanket checking all the attributes and the innerHTML... but I guess we can do something like this.

However, I just realized... you're looping ove rthe `embeds` array, but you're checking `node.attributes` - and you never assign to `node`, so it's still the node ancestor in which we're checking for embeds. We should be checking `embeds[ei]` for these attributes etc.

RadhiFadlillah replied:
> you're looping over the `embeds` array, but you're checking `node.attributes`

:man_facepalming: fixed in [0a23f5](https://github.com/mozilla/readability/pull/526/commits/0a23f536326a5af02b7b10f5771cddfa82385217), now it will check the `embeds` properly.

> I'm not a fan of the joining up of the attributes like this.

I've done a small test and the difference between two regex case is negligible. So, I've decided to use the loop because IMHO it's more readable than joining attributes using map.

RadhiFadlillah commented:
Hi @gijsk, sorry for late response.

As requested, I've fixed several stuff :

- Now it will check the `embeds` properly instead of the `node`.
- I've done a small test and I found that the differences of performance between using regex once on a long string or multiple times on several small strings is almost negligible. So, I decided to use loop for attributes check because IMHO it's more readable than joining attributes using map.
- Now `innerHTML` only checked if embed is an `<object>`.
- Removed variable `viddeoCount`. Now, if the node contains an embedded video, the process will return false immediately and the `node` won't be deleted.
- Since `_clean` and `_cleanConditionally` share the same code, I've updated code in `_clean` following the `_cleanConditionally`.

gijsk commented:
Looks great, thanks!
